Manual paging

krassif opened the pull request:
Hi Jorge,

Been time since the manual-paging in the eachRow. In addition, we now have throttling in stream.
Hopefully, this concludes the manual paging work.

Please review and make these part of the official driver release - tired of using two different driver versions in Peerbelt.

Let me know your thoughts, comments.

Thank you,
-Krassi

--
The manual-paging branch is too far behind and the Pull Request looked terrible with 90 or so files to be considered. Closed it and open this one instead.

Perhaps, you can recreate the branch if we need to work there?!

--
feat: support for manual paging:
	in eachRow via fetchSize option spec, resulting in a nextPage method attached to the resultStats in callback(err, resultStats)
	in stream, implicitly via the nodejs stream implementation, or again via fetchSize option spec

fix: according to https://nodejs.org/api/stream.html#stream_readable_push_chunk_encoding:
	"if push() returns false, then we need to stop reading from source", i.e. this.paused = !this.push(this.buffer.shift());

jorgebay commented:
Looks good! Sorry we left out from previous releases your pr.

I will review your pr now, so it can be merged it into a dev branch with the ticket name and add some tests to it.

jorgebay commented on lib/client.js:
```diff
@@ -361,13 +371,33 @@ Client.prototype.stream = function () {
+      // allows for throttling as per the
+      // default nodejs stream implementation 
+      resultStream._valve( function( ) {
+          try {
```
Could you use 2 spaces for indentation to follow the style of the rest of the project?

jorgebay commented on lib/types/result-stream.js:
```diff
@@ -1,5 +1,7 @@
 var util = require('util');
 var stream = require('stream');
+// var assert = require('assert');
```
Remove

jorgebay commented on lib/types/result-stream.js:
```diff
@@ -19,7 +21,32 @@ ResultStream.prototype._read = function() {
+    this.paused = !this.push(this.buffer.shift());
+  }
+  if ( !this.paused && !this.buffer.length && this._readNext ) {
+    // console.log( "_read", this.paused, this.buffer.length );
```
Remove

jorgebay commented on lib/types/result-stream.js:
```diff
@@ -19,7 +21,32 @@ ResultStream.prototype._read = function() {
+
+/**
+ * Allows for throttling, helping nodejs keep the internal buffers reasonably sized.
+ * @param {function} function that triggers reading the next result chunk
```
`@param` should follow the type and name of the parameter:
```
* @param {function} readNext function that triggers reading the next result chunk
```

Remove the `@constructor` jsdoc?

jorgebay commented on lib/types/result-stream.js:
```diff
@@ -19,7 +21,32 @@ ResultStream.prototype._read = function() {
   }
   while (!this.paused && this.buffer.length > 0) {
-    this.paused = this.push(this.buffer.shift());
+    this.paused = !this.push(this.buffer.shift());
```
Good call!

krassif commented:
Just pushed the formatting changes:
 - removed the commented out code
 - removed the @constructor ( not sure how I left it there anyway; good catch )
 - updated the param jdoc
 - indents are two spaces as the rest of the files (no changes were needed for this)

Can you get the latest? No need for a new pull request, right?

Thanks!

jorgebay commented:
I'm closing this one (master local branch) in favour of #89
